### PR TITLE
Make min and max value for slider configurable

### LINKED ISSUE #6776: Slider does not support minValue and maxValue (like Setpoint)
See the documentation on the openHAB (the open source reference implementation of Eclipse smarthome) regarding the slider and setpoint:
https://www.openhab.org/docs/configuration/sitemaps.html#element-type-setpoint

The example for both widgets uses the example of a temperature changer to change the temperature in the kitchen. While the setpoint allows the user to add a min and max value, when using the slider, the user of the UIs can make a mistake and, e.g., set the temperature to 100° C.

The way to solve this could be to use a setpoint in this case (which is not the only case where this issue applies[1]), however, in comparison of both does the setpoint allow a more exact setting of the target value with the disadvantage of less usability, while the slider allows a more useable way of setting the target value, which might not be as easy to set the exact value again (as the user has to "slide").

In my little opinion, I don't see a reason, why the slider should not allow to have a min and max value, like the setpoint, too. This would allow people to decide, what they want: a way to set the fine-grained value exactly with less usability (e.g. this might be useful for temperatures) or to use a slider when the value needs to be set faster and it doesn't matter if it is exactly a specific value (e.g. when setting the brightness, which is more or less a "feeling" of the user whilde setting it).

[1] Another scenario might be a binding, which exposes the brightness of a lamp as a number going from 0 to 255, instead of a percenttype from 0 to 100. A setpoint would be needed to set the brightness above of 100, which is not really useable compared with a slider (which unfortunately hardcodes to 100 as a max value).

### PR DESCRIPTION
**Pull request by FlorianSW**
Like for the setpoint, the min and max value for a slider can differ based
on the underlying thing that is being controlled, so it shouldn't be
hardcoded and a user should be able to change these values.

The default values, if the parameters are not given, are 0 (minValue) and
100 (maxValue), like they're currently, so that users shouldn't see a
difference when they're not changing their configurations.

Fixes #6776 

Signed-off-by: Florian <florian.schmidt.welzow@t-online.de>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
The slider widget was originally only introduced for Dimmer items and no others.
What does it do for number items right now?
If we want to extend its use to number items, shouldn't we also introduce a "step" parameter as for the Setpoints?

**Comment by FlorianSW** [comment]
> Could you please adjust Classic UI too ?

I looked at it, too, however, I've two questions here:
1. The classic UI renders sliders like setpoints, to there're only two buttons, I'm not sure where to put the min and max values there (no-where?)
2. Where is the non-minified source code of the "Logic.js" where I would expect to put some min/max-value handling?

> What does it do for number items right now?
E.g. the temperature (like in the openHAB example) or, in the actual use case, controlling the brightness of an item, which expects the brightness as a range from 0 to 255 instead of 0 to 100. Currently it would need to be "translated" in a rule to work with a slider.

> If we want to extend its use to number items, shouldn't we also introduce a "step" parameter as for the Setpoints?

Hmm, I don't think so. The `step` parameter in the setpoint means, that when pressing the minus or plus to increase or decrease the current value by the value of step. Whereas I would expect a slider to change the value based on the relative position on the slider. Is it understandable what I mean here? :D

**Comment by maggu2810** [comment]
For a slider you need to know the minimum step size, too. Or how do you want decide which values the slider position should be assigned to?

**Comment by FlorianSW** [comment]
Why? The assignment of the values doesn't change when you make the min and Max values configurable (the slider had min and Max values before, too, they were just hardcoded). The slider had a specific width and based on that width the range between the min andax values are assigned to the slider relatively. Let's say, that the slider width is 100%, then each step of the slider position is 1%, it doesn't really matter what the absolute values are then, they are calculated by the relative position of the slider.

**Comment by maggu2810** [comment]
I did not have a look at the code so I could be wrong and if you already checked it...
I had in mind that a slider has a fixed minimum of 0 a maximum of 100 and a step size of 1.
So, you cannot set e.g. 3,75 using a slider (but I really don't know if this is correct).
My assumption has been, if we could change the minimum and maximum (e.g. min=0 and max=1) we need also to set if the step size should be 1, 0.1, 0.01, 0.001 or so on.
It would not make sense to allow a more granular step size if it is not handled on the item's linked channel.

**Comment by FlorianSW** [comment]
So, for now the slider has a default step size (in basic UI only), which is 1 (see the [mdl documentation](https://getmdl.io/components/#sliders-section)), so it would be possible to add a step parameter, too. I really do not care so much about this part of it, so I'm ok with adding a step parameter, too (which should default to 1 I think).

**Comment by FlorianSW** [comment]
The latest commit changes the slider to have another new parameter (step) and made changes to the basic UI to take this into account.

I didn't changed the classic UI so far because I want to check with you guys first about one thing:
If we make this change to the classic ui (min, max and step), then the widget in the classic UI renders really like the setpoint widget. The only difference now is, that the slider currently uses the "magic" INCREASE and DECREASE command to set the new value, the setpoint pre-calculates the new value, as it needs to take the step parameter into account.

My solution here would be to use the same widget for a slider and a setpoint in the classic UI, is that ok? Any other opinions? :)

**Comment by maggu2810** [comment]
I am not a user of the basic or classic UI, so hopefully another one can comment.

**Comment by kaikreuzer** [comment]
@FlorianSW I do not think that anything should be changed in the Classic UI, because the widget just works fine and min/max do not have any impact on it. Completely changing its behavior to not send INCREASE/DECREASE anymore would rather be a breaking change that might annoy some users.

File: extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/SliderRenderer.java [hunk 1]
```diff
@@ -81,6 +85,27 @@ public boolean canRender(Widget w) {
...
+    }
+
+    private String minValueOf(Slider slider) {
+        if (slider.getMinValue() != null) {
+            return slider.getMinValue().toString();
+        }
+        return "0";
+    }
+
+    private String setpOf(Slider slider) {
```
Comment (kaikreuzer) [inline extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/SliderRenderer.java]: was this meant to read `stepOf`?

Comment (FlorianSW) [reply extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/SliderRenderer.java]: oops, my fault, I'll fix it :)

**Comment by kaikreuzer** [comment]
May I ask you to also do an update to https://www.openhab.org/docs/configuration/sitemaps.html#element-type-slider?

**Comment by FlorianSW** [comment]
Of course, I'll submit a PR :) May I ask if there's already a plan when (not exactly of course, just an estimation) this will be included in an ESH release, as well as an openHAB release?

**Comment by kaikreuzer** [comment]
Will be in the latest openHAB distro snapshots later this week, so the docs can be adapted right away!

**Comment by maniac103** [comment]
Just jumping in here to thank you for not forgetting about step ;-)

> Why? The assignment of the values doesn't change when you make the min and Max values configurable (the slider had min and Max values before, too, they were just hardcoded). The slider had a specific width and based on that width the range between the min andax values are assigned to the slider relatively. Let's say, that the slider width is 100%, then each step of the slider position is 1%, it doesn't really matter what the absolute values are then, they are calculated by the relative position of the slider.

That's all right, but still, step is absolutely needed in order to have a defined granularity. Say the 0..100% slider is 100 pixels wide: the minimum increment then would be 1%. If the slider grows to 500 pixels (presentation is up to the UI after all), the minimum increment suddenly becomes 0.2%. Since the UI doesn't know whether the underlying number should be integer or floating point, what increment between data points should it use? Having 'step' solves that uncertaincy, so thanks for not making my life hard when implementing this in the Android app :+1:

**Pull request merged**